Textual Updates for Mifare Classic Command set

nivrij opened the pull request:
I tried to look at all of the MFC commands to make the help text and error texts easier to understand.
This ranged from small grammatical fixes to multiple new lines of text. This should improve the general readability and understand for the end user.

doegox commented:
Beware you introduced extra spaces & tabs.
Please run a "make style" against your changes.
I just pushed now a make style on master, so if you merge/rebase and do a make style on top it should fix your files.
Thank you.

iceman1001 commented:
Interesting,  nice contribution,

Since the client is used on many different sizes on terminals,  the screen space is sparse.   
Which means we use a non-verbose ,  direct kind of text style in order to keep it short.

And of course since this project has been having 100+ contributors over the years,  there hasn't been a strict policy about it.
As the project grows on to more platforms,  the issue becomes more apparent. 

One of which, is the new helptext   "->"   where you add a comment to the end of the examples.  It has generated very long lines...  A nightmare for ppl on android or small tablets.  

I will go through it and have a think.   It is clear that we need better texts but it has to be with clear guidelines,  to help when updating other text parts.

nivrij commented:
Thank you for your consideration.

I have ran make style as requested which seemed to do as you wanted.

Since there were no guidance on line length or character limits per line, I took the standing that there were none. Thus it was down to my implementation. To this end I did try to make each line concise and to the point, however this was easier said than done.

If you have any specific guidance or ideas for this going forward I would be happy to try and update this PR and future changes to incorporate said guidance. But as it stands currently this PR is the work of no guidance regarding textual updates with only the end users' usability in mind.

The other idea I had to provide more lengthy help was through the use of the 'PrintAndLogEx(HINT)' messages which can be turned on/off via the preferences file. There it would not impact all users by default and can be turned off for those with less screen real estate. However, most of the changes I made in this PR I find would not be appropriate to be hints.


I am open to further suggestions and any changes that need to be made for this PR to be merged.

iceman1001 commented:
the hints system isn't suppose to give helptexts,  its suppose to give a hint.

I will go through this PR and come up with some guide lines,  in order to make all future help text changes uniform.

nivrij commented:
Thank you.
I understand the difference between 'hint' and 'help'. The hint texts I meant would be supplemental information to some of the error messages to help steer the user to correct said error they are facing.
This is purely an idea at present, will of course need more work to make it happen.

iceman1001 commented on client/src/cmdhfmf.c:
```diff
@@ -2570,11 +2573,13 @@ static int CmdHF14AMfAutoPWN(const char *Cmd) {
                   "hf mf fchk --1k -k FFFFFFFFFFFF                --> Key recovery against MIFARE Classic 1k\n"
                   "hf mf fchk --2k -k FFFFFFFFFFFF                --> Key recovery against MIFARE 2k\n"
                   "hf mf fchk --4k -k FFFFFFFFFFFF                --> Key recovery against MIFARE 4k\n"
+                  "hf mf chk --1k -k FFFFFFFFFFFF -k A0A1A2A3A4A5 --> Check all sectors, A&B keys against MIFARE Classic 1k\n"
```
`fchk`

mwalker33 replied:
Just catching up.
Line (approx.) 116 
PrintAndLogEx(FAILED, "This is expected to happen in 25% of all cases. Trying again with a different reader nonce...");

Don't we need %% to escape the % ?

iceman1001 commented on client/src/cmdhfmf.c:
```diff
@@ -3691,18 +3699,18 @@ int CmdHF14AMfELoad(const char *Cmd) {
-                  "hf mf eload -f hf-mf-01020304.bin\n"
-                  "hf mf eload --4k -f hf-mf-01020304.eml\n"
+                  "hf mf eload -f hf-mf-01020304.bin      --> Use the specified file's data to fill 1KB of emulator memory\n"
+                  "hf mf eload --4k -f hf-mf-01020304.eml --> Fill 4KB of emulator memory from the specified file\n"
```
these two description texts feels bad.  
```
--> Fill 1KB of emulator memory from file
--> Fill 4KB of emulator memory from file
```

iceman1001 commented on client/src/cmdhfmf.c:
```diff
@@ -4084,15 +4093,15 @@ static int CmdHF14AMfEKeyPrn(const char *Cmd) {
-                  "hf mf ekeyprn --1k --> print MFC 1K keyset\n"
-                  "hf mf ekeyprn -w   --> write keys to binary file"
+                  "View in use keys from emulator memory",
+                  "hf mf ekeyprn --1k --> View MIFARE 1K keys from emulator memory\n"
```
feels a bit cake on cake  explaining  _keys from emulator memory_  in the description of samples.
```
View all keys
View MIFARE 1K keys
```

iceman1001 commented on client/src/cmdhfmf.c:
```diff
@@ -4187,14 +4196,15 @@ static int CmdHF14AMfCSetUID(const char *Cmd) {
     CLIParserInit(&ctx, "hf mf csetuid",
-                  "Set UID, ATQA, and SAK for magic gen1a card",
+                  "Set UID, ATQA, and SAK for magic gen1a card\n"
+                  "Only works with magic gen1a cards\n",
```
same cake on cake feeling.  First state that its for Gen1a,   then express that once more in the next line

iceman1001 commented on client/src/cmdhfmf.c:
```diff
@@ -4250,8 +4260,8 @@ static int CmdHF14AMfCSetUID(const char *Cmd) {
 
-    PrintAndLogEx(SUCCESS, "Old UID... %s", sprint_hex(old_uid, uidlen));
-    PrintAndLogEx(SUCCESS, "New UID... %s ( %s )",
+    PrintAndLogEx(SUCCESS, "Old UID:  %s", sprint_hex(old_uid, uidlen));
```
No,  don't fiddle with `...`     There is my long going project of moving all":" in the client to dots instead.

iceman1001 commented on client/src/cmdhfmf.c:
```diff
@@ -4261,10 +4271,11 @@ static int CmdHF14AMfCSetUID(const char *Cmd) {
     CLIParserInit(&ctx, "hf mf cwipe",
-                  "Wipe gen1 magic chinese card.\n"
-                  "Set UID / ATQA / SAK / Data / Keys / Access to default values",
+                  "Wipe a gen1 magic card.\n"
```
Consistency with  `Gen1a` vs `gen1`    all across this pr in term of Capital letter or not.

iceman1001 commented on client/src/cmdhfmf.c:
```diff
@@ -4364,9 +4376,9 @@ static int CmdHF14AMfCSetBlk(const char *Cmd) {
-                  "or from emulator memory.",
+    CLIParserInit(&ctx, "hf mf cload\n",
+                  "Load magic gen1a card with data from (bin/eml/json) dump file or from emulator memory.\n"
+                  "Only works with magic gen1a cards",
```
same cake on cake feeling here.   First state its for Gen1a,   then  repeat it next line. 
Here you also see on example where I tried breaking lines up.

iceman1001 commented on client/src/cmdhfmf.c:
```diff
@@ -4625,19 +4638,20 @@ static int CmdHF14AMfCGetSc(const char *Cmd) {
-                  "hf mf csave --4k"
+                  "Save magic gen1a card memory into three files (BIN/EML/JSON) and/or into emulator memory\n"
+                  "Only works with magic gen1a cards",
+                  "hf mf csave                 --> Save a MIFARE 1K card to file using the UID as the filename\n"
```
here the help text should explain the  UID as the filename and not the individual sample descriptions.

iceman1001 commented on client/src/cmdhfmf.c:
```diff
@@ -4625,19 +4638,20 @@ static int CmdHF14AMfCGetSc(const char *Cmd) {
         arg_lit0(NULL, "2k", "MIFARE Classic/Plus 2k"),
         arg_lit0(NULL, "4k", "MIFARE Classic 4k / S70"),
-        arg_lit0(NULL, "emu", "from emulator memory"),
+        arg_lit0(NULL, "emu", "Save to emulator memory"),
```
Nice catch

iceman1001 commented on client/src/cmdhfmf.c:
```diff
@@ -4779,14 +4793,15 @@ static int CmdHF14AMfCView(const char *Cmd) {
     CLIParserInit(&ctx, "hf mf cview",
-                  "View `magic gen1a` card memory",
+                  "View all of the memory in a magic gen1a card\n"
+                  "Only works with magic gen1a cards",
```
same cake on cake

iceman1001 commented on client/src/cmdhfmf.c:
```diff
@@ -5167,9 +5182,9 @@ static int CmdHF14AMfMAD(const char *Cmd) {
-                  "hf mf mad --dch -k ffffffffffff -> decode CardHolder information\n");
+                  "hf mf mad        			--> Shows MAD if it exists\n" //4 spaces, 3 tabs
+                  "hf mf mad --dch -k ffffffffffff 	     --> Decode CardHolder information\n" //6 spaces, 1 tab
+                  "hf mf mad --aid e103 -k ffffffffffff -b  --> Shows NDEF data if it exists. Read card with custom key and key B\n"); //2 spaces
```
remove the confusing end line comments.  and try line up the `-->`  so it will end up same in the client.

iceman1001 commented on client/src/cmdhfmf.c:
```diff
@@ -5324,9 +5339,9 @@ int CmdHFMFNDEFRead(const char *Cmd) {
-                  "hf mf ndefread --aid e103 -k ffffffffffff -b -> shows NDEF data with custom AID, key and with key B\n");
+                  "hf mf ndefread  			--> shows NDEF parsed data\n"
+                  "hf mf ndefread -vv 				--> shows NDEF parsed and raw data\n"
+                  "hf mf ndefread --aid e103 -k ffffffffffff -b  --> shows NDEF data with custom AID, key and with key B\n");
```
align `-->`

iceman1001 commented on client/src/cmdhfmf.c:
```diff
@@ -5930,8 +5945,8 @@ static int CmdHF14AMfView(const char *Cmd) {
-                  "Print a MIFARE Classic dump file (bin/eml/json)",
-                  "hf mf view -f hf-mf-01020304-dump.bin"
+                  "View the contents of a MIFARE Classic dump file (bin/eml/json)",
+                  "hf mf view -f hf-mf-01020304-dump.bin --> View the contents of the supplied bin file"
```
Don't think you need a sample description ...

iceman1001 commented on client/src/mifare/mifarehost.c:
```diff
@@ -871,12 +871,12 @@ int mfCSetUID(uint8_t *uid, uint8_t uidlen, uint8_t *atqa, uint8_t *sak, uint8_t
     int res = mfCGetBlock(0, block0, params);
     if (res == 0) {
-        PrintAndLogEx(SUCCESS, "old block 0... %s", sprint_hex_inrow(block0, sizeof(block0)));
+        PrintAndLogEx(SUCCESS, "Old block 0:  %s", sprint_hex_inrow(block0, sizeof(block0)));
```
don't touch `...`

iceman1001 commented on client/src/mifare/mifarehost.c:
```diff
@@ -1183,20 +1183,20 @@ int detect_classic_nackbug(bool verbose) {
-                PrintAndLogEx(SUCCESS, "num of auth requests  : %u", auths);
-                PrintAndLogEx(SUCCESS, "num of received NACK  : %u", nacks);
+                PrintAndLogEx(SUCCESS, "Num of auth requests:  %u", auths);
+                PrintAndLogEx(SUCCESS, "Num of received NACK:  %u", nacks);
```
the colon's can be replaced with `...`

iceman1001 reviewed (changes requested):
Overall an nice textual improvement,   my comments comes from the fact that you did this on your own not knowing what is going with change process of the client.   How would you know? 

So,  one comment I have is what I call the "cake of cake"  feeling in some of the suggested text revisions.   Where something is explained once and then made overly expressed in a second line.

nivrij commented:
Thank you for taking the time to review the changes I have made, it is greatly appreciated.
I have made some changes based on the comments you shared but have not pushed them yet.

The "cake on cake" is not 'overly expressed', it is but a simple reiteration for all of the c*/gen1a commands. I do not see an issue with leaving the same statement on all of those commands.

I have changed all "Gen1a" texts to be consistently spelled with an uppercase 'G'. Given there is no guidance or agreed format (upper or lowercase g) for this, I made the decision myself to adopt the uppercase G. Should this be an issue, this can be easily changed to accommodate your comments.

I disagree with the use of '...' instead of ':'. 
Ellipsis should not be used in this manner as it suggests pause and infers a dramatic response; good for prose, not software. This is why I suggested using a colon which is more commonly used to detail relevant information outlined before the colon. eg "Here are the things you asked for: bread, milk, kalops, cheese". I also found more use of colons to detail information than ellipsis in the files I have changed which contributed to my reasoning to change them to colons. 

I would like to ask for more elaboration on a point raised here:
"these two description texts feels bad.
--> Fill 1KB of emulator memory from file
--> Fill 4KB of emulator memory from file"
Please detail why these 'feel bad' so the changes do not make the issue worse and actually rectify your issue. 

Regarding the '-->' lines in the sample texts; I found that aligning them did not have the correct spacing when viewing the changes in the client. Therefor I experimented with more spacing and tabs which I was able to align them correctly in the client but not in the source files. However, is this something that is solved by running 'make style'? If not, then the odd spacing/tabs will need to stay in order to align them in the client.

Similarly with how long a line can be before it needs to be broken into a new line. Given there is no guidance on this and I had no issues in my testing (no mobile or tablet testing) I do not feel this is necessary to change. I would be open to adopting any suggestions you may have regarding character limits per line but I feel this would have a much larger affect on the repository as a whole, not just my PR. 

I believe I have addressed all of your comments in some detail. Should I have missed something please make me aware and Ill be happy to address it. I plan to push the changes when we come to an agreement on some of the above points and/or formal guidance is created regarding textual updates.

mwalker33 commented:
Just catching up.
Line (approx.) 116
PrintAndLogEx(FAILED, "This is expected to happen in 25% of all cases. Trying again with a different reader nonce...");

Don't we need %% to escape the %  or is this handled inside PrintAndLogEx (I assume it has been complied and tested)?

iceman1001 commented:
I don't see any pushed changes. 
@mwalker33  yes,  that text is a invalid string format.

nivrij commented:
> I don't see any pushed changes.
Should you read my previous comment in full you would understand why. Please see my previous points and address those before commenting on changes again.

> @mwalker33 yes, that text is a invalid string format.
This was my mistake and has been corrected accordingly. Thanks for spotting it, I never noticed that message when I was testing the changes.

iceman1001 commented:
Ok,  interesting attitude.
